### DISCUSSION TITLE
What happened to ANTLR 3's Symbol Table?

### DISCUSSION THREAD

**Discussion by AllynShell** [antlr/antlr4#4681] [Ideas]
ANTLR 3 could generate a Symbol Table. Why hasn't ANTLR4 brought this feature forward?

I recommend that ANTLR4 restore the scope sensitive Symbol Table to track types, constants, variables, operators and functions and to assure declare before assign and assign before use and validate which operators are valid for which types and what conversions must be inserted to assure correctness.

**Comment by jimidle** [comment]
Symbol table and scopes are not within the purvey of antler. I’m not sure
what you mean when you say ANTLR3 generated this? I seem to remember a post
about doing such things - perhaps you are referring to that?

**Reply by AllynShell** [reply]
See  https://theantlrguy.atlassian.net/wiki/spaces/ANTLR3/pages/2687062/LLVM about 2/3 of the way through this ANTLR3 blog post created by TerenceP on Aug 10, 2009 is the explanation of how to use ANTLR3's Symbol Table.

**Reply by kaby76** [reply]
> See https://theantlrguy.atlassian.net/wiki/spaces/ANTLR3/pages/2687062/LLVM about 2/3 of the way through this ANTLR3 blog post created by TerenceP on Aug 10, 2009 is the explanation of how to use ANTLR3's Symbol Table.

The symbol table used in the [example](https://theantlrguy.atlassian.net/wiki/download/attachments/2687062/llvm.tar.gz?version=1&modificationDate=1259093793000&cacheVersion=1&api=v2) of the [post](https://theantlrguy.atlassian.net/wiki/spaces/ANTLR3/pages/2687062/LLVM#LLVM-Sourcecode) is implemented in the example. It is not generated by the Antlr Tool nor part of the Antlr3 runtime. While the grammar in the example contains actions that manipulate the state of the SymbolTable, the Antlr Tool copies actions to the generated .java. It is not part of Antlr3 in any way.

For Antlr4, the symbol table is implemented in Java, and only Java, at https://github.com/antlr/symtab, but it may not work the same as that provided in the example. I ported [the symbol table code to C#](https://github.com/kaby76/Piggy/tree/86c38412b9228760572953eb7ba3edadaa325563/Symtab), and had to make considerable changes for what I was trying to do. But, yes, it is not part of the Antlr4 runtime.

**Reply by AllynShell** [reply]
OK, it is  ANTLR4's Java implementation. 
Let me change this to a request:
Could a scope sensitive Symbol Table be be provided in ANTLR4's C++ implementation?

**Reply by mike-lischke** [reply]
Unlikely to happen, however in my [antlr4-c3](https://github.com/mike-lischke/antlr4-c3) code completion module I also added a symbol table compatible with ANTLR4. It's written in TypeScript, so it should be pretty straight forward to port that to any language.

This repo also contains ports of the code completion engine for C#, C++ and Java, so if you do a port of the symbol to any of these languages I would be happy if you'd open a PR for including that in the repo.

**Reply by AllynShell** [reply]
I have been thinking about a symbol table for ANTLR4 a lot. Because ANTLR4 is currently the best choice around for lexical analysis and parsing a language under development, it makes a lot of sense to bridge the gap again to LLVM's IR. I would like to see some facility to designate scope within the ANTLR4 grammar. (e.g. /\ to designate the start of a scope and \\/ to designate the end of a scope.)

unscoped_statements : '{' statement* '}' ;
scoped_statements :  '{' /\ declaration* statement* \\/ '}' ;
